sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventDataAggregator.java

```diff
@@ -68,11 +72,25 @@ public void subscribe(CoreSubscriber<? super EventDataBatch> actual) {
         final EventDataAggregatorMain subscription
             = new EventDataAggregatorMain(actual, namespace, options, batchSupplier, partitionId, LOGGER);
 
-        if (!downstreamSubscription.compareAndSet(null, subscription)) {
+        if (DOWNSTREAM_SUBSCRIPTION.compareAndSet(this, null, subscription)) {
+            source.subscribe(subscription);
+        } else {
             throw LOGGER.logThrowableAsError(new IllegalArgumentException("Cannot resubscribe to multiple upstreams."));
         }
+    }
+
+    int getNumberOfEvents() {
+        Subscription downstream = downstreamSubscription;
+        if (downstream == null) {
+            return 0;
+        }
 
-        source.subscribe(subscription);
+        if (!(downstream instanceof EventDataAggregatorMain)) {
+            throw LOGGER.logExceptionAsError(new IllegalStateException(
+                "Subscription is not of type EventDataAggregatorMain. Actual: " + downstream.getClass()));
+        }
+
+        return ((EventDataAggregatorMain) downstream).getNumberOfEventsInCurrentBatch();
```

Comment:
what I meant in the last comment was was - with `downstreamSubscription` is of type `EventDataAggregatorMain`, we can simplify this function -

```java
int getNumberOfEvents() {
  if (downstreamSubscription == null) {
    return 0;
   }
   return downstreamSubscription.getNumberOfEventsInCurrentBatch();
}

Reply:
I think this commit addresses that. I had missed it initially. https://github.com/Azure/azure-sdk-for-java/pull/44392/commits/ce3560adabd00c21031e8263a047e012626cc301